### DISCUSSION TITLE
How to  find the spectral coherence between the columns of a BIG data matrix/2d array by using python mpi4py - Parallel computation?

### DISCUSSION THREAD

**Discussion by ssalam21** [mpi4py/mpi4py#153] [Q&A]
I've a big matrix/2D array for which every possible column-pair I need to find the coherence by parallel computation in python (e.g. mpi4py). Coherence [a function] are computed at various child processes and the child process should send the coherence value to the parent process that gather the coherence value as a list. To do this, I've created a small matrix and list of all possible column pairs as follows:

`import numpy as np
from scipy import signal
from itertools import combinations
from mpi4py import MPI

comm = MPI.COMM_WORLD
nproc = comm.Get_size()
rank = comm.Get_rank()

data=np.arange(20).reshape(5, 4)
#List of all possible column pairs
data_col = list(combinations(np.transpose(data), 2)) #list

# Function creation
def myFunc(X,Y):
..................
..................
    return Real_coh

if rank==0:
    Data= comm.scatter(data_col,root=0) #col_pair`

Can anyone suggest me how to proceed further. You are welcome to ask any questions/clarifications. Expecting your cordial help. Thanks

**Comment by ssalam21** [comment]
I  try to move with the following script but fail to get the output. Don't identify where is the problem. 

`import numpy as np
from scipy import signal
from mpi4py import MPI
import h5py as t

chunk_len = 5000 # No. of rows of a matrix
num_c = 34    # No. of column of the matrix

# Actual Dataset
data_mat = np.random.random((10000, num_c))

shape = (chunk_len, data_mat.shape[1])
chunk_size = (chunk_len, 1)
no_of_chunks = data_mat.shape[1]

with t.File('file_name.h5', 'w') as hf:
    hf.create_dataset("chunked_arr",  data=data_mat, chunks=chunk_size, compression='lzf')
del data_mat

def myFunc(dset_X, dset_Y):
    ..............
    ............
    return Real_coh

res = np.zeros((num_c, num_c))

comm = MPI.COMM_WORLD
size = comm.Get_size()
rank = comm.Get_rank()

for i in range(num_c):
    with t.File('file_name.h5', 'r', libver='latest') as hf:
        dset_X = hf['chunked_arr'][:, i]  # Chunk data reading
    if i % size == rank:
        for j in range(num_c):
            with t.File('file_name.h5', 'r', libver='latest') as hf:
                dset_Y = hf['chunked_arr'][:, j] # Chunk data reading
            res[i][j] = spac(dset_X, dset_Y)
comm.Barrier()
print('Shape of final result :', res.shape )`

**Comment by dalcinl** [comment]
1. If you do not have experience with MPI in general and mpi4py in particular, I would suggest you start by making your code work using Python's `concurrent.futures` module. Next move to use `mpi4py.futures`. I you are still not happy with the performance and/or scalability, then (and only then) you should try to look for a more ad-hoc solution.
2. You posted this question to mpi4py's mailing list. Posting it here as an issue (that I converted as a discussion) will not help you to get a faster answer.
3. You should do a better job formatting your question for readability.
4. Sending a private email after a) posting to the mailing list, and b) open a GitHub issue, is really unnecessary.

**Reply by ssalam21** [reply]
Let's say, I've a BIG matrix/ 2D array (Matrix size is ~15GB, so I write the file in HDF5 format and want to load a pair of 2 columns, calculate the coherence, store the result and move for another pair of columns, and so on). I want to find the coherence between every possible columns. So, I want to transfer every possible pair of columns to the respective processor of  the cluster and after doing the computation, I want to gather all the results to the parent processor (at rank=0). As I'm new to use mpi4py, so I can't proceed how to send the pair of columns to the different processors. 

One more question, is there any easier algorithm to do this job? Any suggestion/solution is highly appreciated.

**Reply by dalcinl** [reply]
As I said before, learn how to implement your algorithm using `concurrent.futures` and a small test matrix using your laptop or workstation. Once you have it working, you can use `mpi4py.futures` with very little effort, and you can scale it up to a large number of compute nodes and larger matrix sizes.
